modin/core/execution/ray/common/engine_wrapper.py

```diff
@@ -47,6 +48,8 @@ def _deploy_ray_func(func, *args, **kwargs):  # pragma: no cover
 class RayWrapper:
     """Mixin that provides means of running functions remotely and getting local results."""
 
+    _func_cache = {}
```

Comment:
Please add a type hint here.

Have you estimated the cache size that might result?

Reply:
No, but my assumption is that it must not be greater than pandas.DataFrame.* + a limited set of internal modin functions. If it's not true, than something is wrong and should be fixed. My initial approach was to use weak references, but it could be error-prone and results in performance issues, when short living objects are added to the cache and never reused. It's easier to detect such issues, when using strong references.
I think, it would be reasonable to add a size assert here, but we need to roughly estimate the maximum cache size.

Reply:
> I think, it would be reasonable to add a size assert here, but we need to roughly estimate the maximum cache size.

It’s not easy to make an estimate by bytes; it would be nice to have at least the maximum number of keys.

By the way, doesn't Ray itself cache anything?

Does it make sense to do caching for other engines now? (Dask and Unidist)

Reply:
Perhaps, it does cache, but i'm not sure, because ray.put() always returns a new object reference:

```python
>>> id(ray.put(pandas.DataFrame.add))
140361903189728
>>> id(ray.put(pandas.DataFrame.add))
140361903189584
>>> id(ray.put(pandas.DataFrame.add))
140360098017280
```

Reply:
We could either add something like:
```python
assert len(_func_cache) < 64
```
or use lru cache here, but I'm not sure if it makes sense.

Reply:
> We could either add something like:
assert len(_func_cache) < 64

Maybe we’ll make 128 functions and it won’t crash with an error if the 129th function appears, but we’ll delete the very first of the remaining functions and add a new one instead?

Reply:
We can just use lru cache with max 128 items, but it hides the problem when the function filter does not work properly and we add/remove short-living functions to the cache. In case of explicit  size check, if we hit the assertion error, then either the limit should be increased or the filter is broken and should be fixed.

Reply:
> In case of explicit size check, if we hit the assertion error, then either the limit should be increased or the filter is broken and should be fixed.

In this case, it is the users who will most likely encounter errors, not us. It turns out that something that worked before may break due to caching; this is an unacceptable option.

Reply:
That's why we must either properly estimate the limit, or set a large limit, for example, 1024. In case of an invalid filtering, we will definitely hit this limit in unit tests.
Another option - use warning instead of assert.

Reply:
1024 limit and using warnings LGTM.

> Maybe we’ll make 128 functions and it won’t crash with an error if the 129th function appears, but we’ll delete the very first of the remaining functions and add a new one instead?

>> but it hides the problem when the function filter does not work properly and we add/remove short-living functions to the cache. 

We can introduce a separate counter that will count how many functions have been cached.

`pytest` seems to have a function that is called at the end of the testing session, we can check how many functions were cached during the tests, however, we do not run all tests with one `pytest` call, so such a check is not sufficient. There is a feeling that such things can mainly be caught using complex integration tests, which we do not have.

Reply:
Added the size check.